src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -150,34 +197,47 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
-            {
-                let ty = ty_to_string(ty);
-                (format!(" for `{}`", ty),
-                 format!("the explicit type `{}`, with the type parameters specified", ty))
-            }
-            Some(ty) if &ty.to_string() != "_" &&
-                ty.to_string() != name &&
-                // FIXME: Remove this check after `impl_trait_in_bindings` is stabilized. #63527
-                (!ty.is_impl_trait() || self.tcx.features().impl_trait_in_bindings) &&
-                !ty.is_closure() => // The suggestion doesn't make sense for closures.
-            {
-                let ty = ty_to_string(ty);
-                (format!(" for `{}`", ty),
-                 format!(
-                     "the explicit type `{}`, where the type parameter `{}` is specified",
-                    ty,
-                    name,
-                 ))
-            }
+        let mut err = struct_span_err!(
+            self.tcx.sess,
+            err_span,
+            E0282,
+            "type annotations needed{}",
+            ty_msg,
+        );
+
+        let suffix = match local_visitor.found_ty {
             Some(ty::TyS { sty: ty::Closure(def_id, substs), .. }) => {
-                let msg = " for the closure".to_string();
                 let fn_sig = substs.closure_sig(*def_id, self.tcx);
+                let ret = fn_sig.output().skip_binder().to_string();
+
+                if let Some(ExprKind::Closure(_, decl, body_id, ..)) = local_visitor.found_closure {
+                    let (arrow, post) = match decl.output {
+                        FunctionRetTy::DefaultReturn(_) => ("-> ", " "),
+                        _ => ("", ""),
+                    };
+                    if let Some(body) = self.tcx.hir().krate().bodies.get(body_id) {
+                        let suggestion = match body.value.node {
+                            ExprKind::Block(..) => {
+                                vec![(decl.output.span(), format!("{}{}{}", arrow, ret, post))]
+                            }
+                            _ => {
+                                vec![
+                                    (decl.output.span(), format!("{}{}{}{{ ", arrow, ret, post)),
+                                    (body.value.span.shrink_to_hi(), " }".to_string()),
+                                ]
+                            }
+                        };
+                        err.multipart_suggestion(
+                            "give this closure an explicit return type without `_` placeholders",
+                            suggestion,
+                            Applicability::HasPlaceholders,
+                        );
+                        err.span_label(span, InferCtxt::missing_type_msg(&name));
+                        return err;
+                    }
+                }
+
+                // This shouldn't be reachable, but just in case we leave a reasonable fallback.
```

Comment:
I'd prefer to leave this in instead of ICEing because I lack the imagination to trigger this case.